db/compaction/compaction_iterator.cc

```diff
@@ -987,8 +1004,12 @@ void CompactionIterator::PrepareOutput() {
         ikeyNotNeededForIncrementalSnapshot() && bottommost_level_ &&
         DefinitelyInSnapshot(ikey_.sequence, earliest_snapshot_) &&
         ikey_.type != kTypeMerge) {
-      assert(ikey_.type != kTypeDeletion && ikey_.type != kTypeSingleDeletion);
-      if (ikey_.type == kTypeDeletion || ikey_.type == kTypeSingleDeletion) {
+      assert(ikey_.type != kTypeDeletion);
+      assert(ikey_.type != kTypeSingleDeletion ||
+             (timestamp_size_ || full_history_ts_low_));
```

Comment:
This function expects no key of type `kTypeDeletion` or `kTypeSingleDeletion`. But single deletion with user defined timestamp can go through this. That's why I added the conditions together to distinguish with  kTypeSingleDeletion and kTypeSingleDeletion with user defined timestamp else it will fail.